File: services/ts-koa-superstruct/v2/echo_service.ts
import * as models from './models'

export interface EchoBodyParams {
    body: models.Message,
}

export type EchoBodyResponse =
    | { status: "ok", data: models.Message }

export interface EchoService {
    echoBody(params: EchoBodyParams): Promise<EchoBodyResponse>
}

File: services/scala-play/v2/services/echo/EchoService.scala
package v2.services.echo

import com.google.inject.ImplementedBy
import scala.concurrent.Future
import v2.models._

@ImplementedBy(classOf[v2.services.EchoService])
trait IEchoService {
  def echoBody(body: Message): Future[EchoBodyResponse]
}

sealed trait EchoBodyResponse
object EchoBodyResponse {
  case class Ok(body: Message) extends EchoBodyResponse
}
